### PR TITLE
Add support for gerrit InRepoConfig with fallback

### PR DESCRIPTION
**Pull request by mpherman2**
PR enables InRepoConfig for Gerrit. If Getting InRepoConfig fails, this will log a warning and fall back to just getting static presubmits. Doing this as we do not have robust gerrit integration tests yet, and I do not want to break Gerrit component.
/assign @chaodaiG @cjwagner 
/hold

### CODE REVIEW

File: prow/gerrit/adapter/adapter.go [hunk 1]
```diff
@@ -310,11 +313,30 @@ func (c *Controller) processChange(logger logrus.FieldLogger, instance string, c
...
-		// TODO: Do we want to add support for dynamic postsubmits?
-		postsubmits := c.config().PostsubmitsStatic[cloneURI.String()]
-		postsubmits = append(postsubmits, c.config().PostsubmitsStatic[cloneURI.Host+"/"+cloneURI.Path]...)
+		postsubmits := []config.Postsubmit{}
+		if gc != nil {
+			postsubmits, err = c.config().GetPostsubmits(gc, cloneURI.Host+"/"+cloneURI.Path, func() (string, error) { return baseSHA, nil })
+			if err != nil {
+				//TODO(mpherman): Return Error once we know this usually works
+				logger.Warn("Failed to get InRepoConfig for Postsubmits")
+				postsubmits = append(postsubmits, c.config().PostsubmitsStatic[cloneURI.Host+"/"+cloneURI.Path]...)
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter.go:336]: moving this line to be under `gc != nil` statement could cause regression for prow instances that don't have `gc` set, this should also be handled as `else` condition

File: prow/gerrit/adapter/adapter.go [hunk 2]
```diff
@@ -310,11 +313,30 @@ func (c *Controller) processChange(logger logrus.FieldLogger, instance string, c
 
 	changedFiles := listChangedFiles(change)
 
+	//Attempt to Get InRepoConfig
+	opts := git.ClientFactoryOpts{
+		CloneURI:       cloneURI.String(),
+		Host:           cloneURI.Host,
+		CookieFilePath: c.cookieFilePath,
+	}
+	gc, err := git.NewClientFactory(opts.Apply)
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter.go:322]: while we don't have integration test, I think we'll need to set up unit test for this, you can use the same local git at https://github.com/kubernetes/test-infra/blob/e84d71dcb03994718ffe9e7d637acf0ae080b0b2/prow/config/inrepoconfig_test.go#L559 for unit testing this

File: prow/gerrit/adapter/adapter.go (hunk 1 above)
Comment (mpherman2) [reply prow/gerrit/adapter/adapter.go:336]: Fixed

File: prow/gerrit/adapter/adapter.go (hunk 2 above)
Comment (mpherman2) [reply prow/gerrit/adapter/adapter.go]: I thought I would not be able to at first because the code to get the prow yaml is too complicated to include in the Unit test for the adapter, but I forgot that the ProwYamlGetter can be mocked. Went back and added unit tests

File: prow/gerrit/adapter/adapter_test.go [hunk 3]
```diff
@@ -1111,8 +1208,25 @@ func TestProcessChange(t *testing.T) {
 				gc:            &gc,
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter_test.go:1208]: I think we should also test scenario where `gc = nil`

File: prow/gerrit/adapter/adapter_test.go [hunk 4]
```diff
@@ -1111,8 +1208,25 @@ func TestProcessChange(t *testing.T) {
...
 				tracker:       &fakeSync{val: fakeLastSync},
 			}
+			cloneURI, err := makeCloneURI(tc.instance, tc.change.Project)
+			if err != nil {
+				t.Errorf("error making CloneURI %v", err)
+			}
+
+			lg, cf, err := localgit.NewV2()
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter_test.go:1216]: since ProwYamlGetter is mocked, seems like this is unused?

File: prow/gerrit/adapter/adapter_test.go (hunk 3 above)
Comment (mpherman2) [reply prow/gerrit/adapter/adapter_test.go:1228]: Done

File: prow/gerrit/adapter/adapter_test.go (hunk 4 above)
Comment (mpherman2) [reply prow/gerrit/adapter/adapter_test.go]: ProcessChange takes a clientFactory, and will only get to InRepoConfig if one is provided (if it is nil it will skip InRepoConfig). This is the best way to create one that is non-nil for the unit tests.

Comment (chaodaiG) [reply prow/gerrit/adapter/adapter_test.go]: would be nice to add a comment explaining it

Comment (mpherman2) [reply prow/gerrit/adapter/adapter_test.go]: Done

File: prow/gerrit/adapter/adapter_test.go [hunk 5]
```diff
@@ -970,6 +1023,63 @@ func TestProcessChange(t *testing.T) {
...
+				Revisions: map[string]gerrit.RevisionInfo{
+					"1": {
+						Number: 1001,
+					},
+				},
+			},
+			instancesMap:     map[string]*gerrit.AccountInfo{testInstance: {AccountID: 42}},
+			instance:         testInstance,
+			shouldError:      false,
+			numPJ:            0,
```
Comment (chaodaiG) [inline prow/gerrit/adapter/adapter_test.go:1080]: there should also be a unit test fetching static jobs when gc is nil

Comment (mpherman2) [reply prow/gerrit/adapter/adapter_test.go:1080]: Done

**Review by chaodaiG — APPROVED** [review]

**Comment by mpherman2** [comment]
This was a good first pass, but I was reflecting on this and I am going to close this pull request and work on getting the cached version of InRepoConfig working for Gerrit instead of the uncached version. Closing this PR.
